src/tests/tracing/eventpipe/userevents/userevents.cs

```diff
@@ -0,0 +1,120 @@
+            string traceFilePath = Path.Combine(appBaseDir, trace);
+
+            if (!File.Exists(recordTracePath) || !File.Exists(scriptFilePath))
+            {
+                Console.WriteLine("record-trace or dotnet-common.script not found. Test cannot run.");
+                return -1;
+            }
+
+            ProcessStartInfo traceeStartInfo = new();
+            traceeStartInfo.FileName = Process.GetCurrentProcess().MainModule.FileName;
+            traceeStartInfo.Arguments = $"{typeof(UserEventsTest).Assembly.Location} tracee";
+            traceeStartInfo.WorkingDirectory = appBaseDir;
+
+            ProcessStartInfo recordTraceStartInfo = new();
+            recordTraceStartInfo.FileName = recordTracePath;
+            recordTraceStartInfo.Arguments = $"--script-file {scriptFilePath}";
+            recordTraceStartInfo.WorkingDirectory = appBaseDir;
+            recordTraceStartInfo.RedirectStandardOutput = true;
+            recordTraceStartInfo.RedirectStandardError = true;
+
+            using Process traceeProcess = Process.Start(traceeStartInfo);
+            using Process recordTraceProcess = Process.Start(recordTraceStartInfo);
+            recordTraceProcess.OutputDataReceived += (_, args) => Console.WriteLine($"[record-trace] {args.Data}");
+            recordTraceProcess.BeginOutputReadLine();
+            recordTraceProcess.ErrorDataReceived += (_, args) => Console.Error.WriteLine($"[record-trace] {args.Data}");
+            recordTraceProcess.BeginErrorReadLine();
+
+            if (!traceeProcess.HasExited && !traceeProcess.WaitForExit(15000))
+            {
+                traceeProcess.Kill();
+            }
+
+            // Until record-trace supports duration, the only way to stop it is to send SIGINT (ctrl+c)
+            kill(recordTraceProcess.Id, SIGINT);
+            if (!recordTraceProcess.HasExited && !recordTraceProcess.WaitForExit(20000))
+            {
+                // record-trace needs to stop gracefully to generate the trace file
+                recordTraceProcess.Kill();
+            }
+
+            if (!File.Exists(traceFilePath))
+            {
+                Console.Error.WriteLine($"Expected trace file not found at `{traceFilePath}`");
+                return -1;
+            }
+
+            if (!ValidateTraceeEvents(traceFilePath))
+            {
+                Console.Error.WriteLine($"Trace file `{traceFilePath}` does not contain expected events.");
+                return -1;
+            }
+
+            return 100;
+        }
+
+        private static bool ValidateTraceeEvents(string traceFilePath)
+        {
+            string etlxPath = TraceLog.CreateFromEventPipeDataFile(traceFilePath);
+            using TraceLog log = new(etlxPath);
+            using TraceLogEventSource source = log.Events.GetSource();
+            bool startEventFound = false;
+            bool stopEventFound = false;
+
+            source.AllEvents += (TraceEvent e) =>
```

Comment:
Is the plan to add more tests here that checks for metadata/fields, rundown events, callstacks etc or should we add some basic verification to this test or plan to extend existing EventPipe tests to also work over UserEvents?

Reply:
This was mainly to add a basic verification that the end to end runtime side (from accepting the ipc message to writing to the tracepoints worked). Since User_events is built on EventPipe, my initial thought is that duplicating the existing eventpipe tests for user events wouldn't be adding anything. I think we can add more tests later on, but not sure what coverage is good and reasonable. I'm not even sure yet if our CI machines have user_events, or if they run with elevated privileges, so this was mainly to see if we can have a basic E2E test going.

Reply:
We should definitely not duplicate but maybe look on extending the existing event pipe tests to run over user events + additional validate logic but agree that is something we could look at later.

So, if this is mainly a smoke test then maybe we should make sure we at least hit things we know is handled in the one-collect library, during the work we hit a number of things that needed special attention, like activity id's, custom metadata and potential stack traces. Right now, these only tests one runtime start/stop event fired under very unique circumstances. Maybe we should do some short multi-threading scenario as well, making sure we won't hit any races in the code path unique to user events?